addons/subscription/models/subscription.py

```diff
@@ -0,0 +1,123 @@
+    note = fields.Text(string='Notes', help="Description or Summary of Subscription")
+
+    @api.model
+    def _auto_end(self):
+        super(Subscription, self)._auto_end()
+        # drop the FK from subscription to ir.cron, as it would cause deadlocks
+        # during cron job execution. When model_copy() tries to write() on the subscription,
+        # it has to wait for an ExclusiveLock on the cron job record, but the latter
+        # is locked by the cron system for the duration of the job!
+        # FIXME: the subscription module should be reviewed to simplify the scheduling process
+        #        and to use a unique cron job for all subscriptions, so that it never needs to
+        #        be updated during its execution.
+        self.env.cr.execute("ALTER TABLE %s DROP CONSTRAINT %s" % (self._table, '%s_cron_id_fkey' % self._table))
+
+    @api.multi
+    def set_process(self):
+        for subscription in self:
+            mapping = {'name':'name', 'interval_number':'interval_number', 'interval_type':'interval_type', 'exec_init':'numbercall', 'date_init':'nextcall'}
+            res = {'model':'subscription.subscription', 'args': repr([[subscription.id]]), 'function':'model_copy', 'priority':6, 'user_id':subscription.user_id.id}
+            for key,value in mapping.items():
+                res[value] = subscription[key]
+            cron = self.env['ir.cron'].create(res)
+            subscription.write({'cron_id':cron.id, 'state':'running'})
+
+    @api.multi
+    def model_copy(self):
+        document_model = self.env['subscription.document']
+        history_model = self.env['subscription.subscription.history']
+        for subscription in self.filtered(lambda sub: sub.cron_id):
+            try:
+                docsource_id = subscription.doc_source.id
+                model_name = subscription.doc_source._name
+            except:
+                raise UserError(_('Please provide another source document.\nThis one does not exist!'))
+
+            docsource_data = {'state':'draft'}
+
+            datefields = document_model.search([('model_id.model', '=', model_name)], limit=1).field_ids.filtered(lambda field : field.value == 'date')
+            fieldnames = datefields.mapped('field_id').mapped('name')
+            docsource_data.update(dict(zip(fieldnames, [fields.Date.context_today(self)] * len(fieldnames))))
+
+            # if there was only one remaining document to generate
+            # the subscription is over and we mark it as being done
+            if subscription.cron_id.numbercall == 1:
+                subscription.write({'state':'done'})
+            else:
+                subscription.write({'state':'running'})
+            self.env[model_name].browse(docsource_id).copy(docsource_data)
+            history_model.create({'subscription_id': subscription.id, 'history_date':fields.Datetime.now(), 'document_id': subscription.doc_source})
```

Comment:
'document_id' is a reference field and will require this particular format of ``model_name,record_id`` in db. So we will need to give value like : ``'%s,%s' % (subscription.doc_source._name,subscription.doc_source.id)`` otherwise it will give trace-back while accessing record of 'subscription.subscription.history'.